### PR TITLE
Dynamic scheduling based on plugins

### PR DESCRIPTION
**Pull request by nateyazdani**
This change allows statscache plugins to specify the exact frequency at which they fire and then generates at load-time the producers needed, which requires some corresponding [changes to statscache_plugins](https://github.com/fedora-infra/statscache_plugins/pull/6). In aim of this goal, I made the following modifications:
- Available plugin classes are loaded at start-up and are initialized by the specific producer to which they are attached (in contrast to how they are currently loaded and initialized all at once)
  - Similar to moksha, generators/iterables are now valid plugin entrypoints. In the future, I would like to add a utility function/class to automate generating a plugin for several frequencies, like how the "volume, by [stuff]" plugins do
- The cache buckets are taken out of `StatsConsumer` and placed in their respective producer objects. This removes the potential issue of needing the producers to all modify the consumer object.
- The `statscache.schdule.Schedule` and `statscache.schdule.Frequency` classes are replaced with a `statscache.plugins.Frequency` class (subjectively relocated because it's part of the plugin API). The new class has a slightly different purpose than it's predecessors. It's an extension of `datetime.timedelta`, and it gets much of the prior functionality from that (e.g., adding/subtracting for `.next()` \ `.prev()`) as well as some additional helpful properties, in particular hashability. All it was missing from the previous implementations was a similarly nice `str()` implementation, which has been added.
  - For this to work, the `BasePlugin.handle()` needed to have the timestamp of the cache empty-ing. A parameter for such a value was already present in the abstract method signature, so I assume that this was already intended.
  - Note that I was not 100% clear on what was happening in the `statscache.schedule` module, so it's entirely possible that I've misinterpreted some of what it does. Please don't hesitate to call me out :)
- There is also some assorted minor improvements that I made (e.g., `idx` -> `ident`) that I felt improved readability, but these are subjective and reversible if preferable.

Finally, I wanted to mention a difference in plugin scheduling behavior. Previously, all plugins were getting fired from each producer; if you look at the initialization output, all 7 default plugins were loaded (and fired) by each producer, meaning all plugins were being run on all frequencies. I don't think that this was the intention, especially since the message volume plugins had specialized versions for particular frequencies. However, if we would like to preserve this sort of behavior, where some or all plugins are run on multiple frequencies, that is something I can easily add.

### CODE REVIEW

File: statscache/producer.py [hunk 1]
```diff
@@ -13,81 +14,94 @@
...
         super(StatsProducerBase, self).__init__(hub)
 
         fedmsg.meta.make_processors(**self.hub.config)
 
-        # Find and save the StatsConsumer instance already created by the hub.
-        # We are going to re-use its backends and db session.
-        self.sister = statscache.utils.find_stats_consumer(self.hub)
+        # set up the cache and connect to the consumer
+        self.cache = []
+        statscache.utils.find_stats_consumer(self.hub).register(self)
```
Comment (ralphbean) [inline statscache/producer.py:29]: Nice, I like this approach.

**Comment by nateyazdani** [comment]
The latest [commit](https://github.com/yazman/statscache/commit/dbe4835f3e27b5e0ca4df4c4d120f6ec506863bd) enhances the statscache.plugins.Frequency class such that it now synchronizes on UTC midnight, while preserving hashability.

**Comment by ralphbean** [comment]
> The latest commit enhances the statscache.plugins.Frequency class such that it now synchronizes on UTC midnight, while preserving hashability.

It looks good to me after a first read through.  I wish I'd seen it earlier, but there's an old test suite in the tests/ dir for the old schedule stuff.  Could you update it for the new Frequency class for peace of mind and future-proofery?

**Comment by nateyazdani** [comment]
@ralphbean done! Note that some of the test cases were specific to the old classes (that allowed irregular frequencies) and were removed entirely, and there are also some that I've added to try to cover the workings of the new class more thoroughly.

As discussed on IRC, synchronizing the longer frequencies to some extent (e.g., ensuring that a daily frequency always fires at 00:00 UTC) is a desirable behavior and has also been incorporated. As part of that, the way the plugin API interacts with this `Frequency` class is slightly changed. The end result is that plugin classes now include an `interval` attribute, which is a `datetime.timedelta` used by `statscache` proper to instantiate a `Frequency` class, which is passed along to the plugins.

**Comment by ralphbean** [comment]
OK, that one default argument needs to be fixed and then this can be merged.  No need to wait for subsequent approval, you can just do it.

**Comment by nateyazdani** [comment]
Ah, I knew something felt wrong when I wrote that, but I just couldn't put my finger on it. This also explains why some calls from the unit tests didn't seem to be affected by `freezegun`. Thanks!

**Pull request merged**